apps/api_web/lib/api_web/event_stream/canary.ex

```diff
@@ -9,19 +9,33 @@ defmodule ApiWeb.EventStream.Canary do
 
   @default_notify_fn &ApiWeb.EventStream.Supervisor.terminate_servers/0
 
-  @spec start_link((() -> any)) :: GenServer.on_start()
-  def start_link(notify_fn \\ @default_notify_fn) do
+  @spec start_link(keyword()) :: GenServer.on_start()
+  def start_link(args \\ []) do
+    notify_fn =
+      case Keyword.fetch(args, :notify_fn) do
+        {:ok, fun} when is_function(fun, 0) ->
+          fun
+
+        {:ok, not_fun} ->
+          {:stop, "expect function/0 for notify_fn, got #{inspect(not_fun)}"}
+
+        _ ->
+          @default_notify_fn
+      end
+
     GenServer.start_link(__MODULE__, notify_fn, [])
   end
 
   @impl true
+  def init({:stop, _} = stop_reason), do: stop_reason
+
   def init(notify_fn) do
     Process.flag(:trap_exit, true)
     {:ok, notify_fn}
   end
 
   @impl true
-  def terminate(:shutdown, notify_fn), do: notify_fn.()
+  def terminate(:shutdown, notify_fn) when is_function(notify_fn), do: notify_fn.()
   def terminate({:shutdown, _reason}, notify_fn), do: notify_fn.()
```

Comment:
If line 30 is never called, couldn't we delete it? It feels wrong to leave in a line that represents a logic error, since we know that `notify_fn` is not a function..